### DISCUSSION TITLE
SQLAlchemy asyncio appears to block the event loop on ASGI web frameworks like FastAPI

### DISCUSSION THREAD

**Discussion by msukmanowsky** [sqlalchemy/sqlalchemy#11804] [Usage Questions]
Hi there,

First off, wanted to thank the maintainers of SQLAlchemy for software that we (and others) use to power our businesses and pet projects every day. Very much appreciate y'all ❤️ .

We have a typical FastAPI and SQLAlchemy setup and we've recently noticed that queries that use `AsyncSession`s with a backend like `aiosqlite` or `asyncpg` actually end up blocking the asyncio event loop which results in an overall slow down of all requests.

I put together [a repo](https://github.com/msukmanowsky/fastapi-sqlalchemy-slowness) that shows how to reproduce this issue alongside a README that shows the results of what we're seeing in our production environments.

To summarize, calls to slow queries `await db_session.execute(some_slow_query)` end up clearly blocking the entire event loop and leading to slow responses to even trivial endpoints like a simple `/ping`.

Here's a sample output from one of our tests showing that while making 100 requests to an endpoint using `AsyncSession`, we actually saw a massive increase in latency to a trivial async endpoint `/ping` (repo contains more context):

```txt
Running test with 100 requests each to paths: ['/count-todos-async', '/ping']
/count-todos-async:
Min: 9.70
Mean: 12.18
Max: 12.69
/ping:
Min: 9.70
Mean: 12.40
Max: 12.61
------ END ------
```

My hypothesis here is that SQLAlchemy's use of `greenlet` for async support doesn't actually play well with one or more of `asyncio` or `anyio`, both of which FastAPI use.

Open to any thoughts y'all have on this!

Mike

**Comment by zzzeek** [comment]
hi -

Opening up a whole repo and debugging your program is a big job, I can show you very simple examples that illustrate that greenlet is not blocking the event loop, would that be helpful?

**Reply by zzzeek** [reply]
I would try running these tests with PostgreSQL or MySQL, and not sqlite.  the aiosqlite driver uses threads itself, and the sqlite3 python driver uses sqlite3 in process.   Meaning, you aren't doing any real asyncio with SQLite, it's all file access, not network access, which is not actually concurrent.

**Reply by msukmanowsky** [reply]
Should've clarified, the repo is _very_ stripped down FastAPI example (tried to keep it as lightweight as I could). [This is the route](https://github.com/msukmanowsky/fastapi-sqlalchemy-slowness/blob/main/src/fastapi_sqlalchemy_slowness/main.py#L74-L83) that demonstrates us using a [simple AsyncSession dependency](https://github.com/msukmanowsky/fastapi-sqlalchemy-slowness/blob/main/src/fastapi_sqlalchemy_slowness/dependencies.py#L9-L14) to perform a taxing query (group by count over 1m rows)

Would appreciate an example showing SQLAlchemy doesn't block the asyncio event loop in anyway as that may spur me to open a separate issue with the FastAPI team.

**Reply by msukmanowsky** [reply]
> I would try running these tests with PostgreSQL or MySQL, and not sqlite. the aiosqlite driver uses threads itself, and the sqlite3 python driver uses sqlite3 in process. Meaning, you aren't doing any real asyncio with SQLite, it's all file access, not network access, which is not actually concurrent.

We see the exact same performance with PostgreSQL and `asyncpg` (this is what we run in prod which was the motivation for looking into this in the first place). The example also isn't so much about testing truly async queries to the db, but demonstrating that any long running async query seems to slow all FastAPI endpoints (which points to the event loop being stalled).

**Reply by zzzeek** [reply]
Here's a program that uses SQLite in a synchronous fashion so that all database calls block the event loop.  To make sure it illustrates the event loop being blocked, a few of the requests do an `asyncio.sleep()` instead.

```py
import os
import asyncio
from sqlalchemy import text
from sqlalchemy import create_engine

async def setup(engine):
    with engine.connect() as conn:
        conn.exec_driver_sql("""
create table big_table (
    some_col int
)
        """)
        conn.exec_driver_sql(
            "insert into big_table (some_col) values (?)",
            [(i % 100, ) for i in range(1000000)]
        )
        conn.commit()


async def work(engine, i):
    if i in (2, 5):
        print(f"starting sleep {i}")
        await asyncio.sleep(.1)
        print(f"ending sleep {i}")
        return

    with engine.connect() as conn:
        print(f"step1 {i}")
        result = conn.execute(text("select count(*) from big_table group by some_col"))
        print(f"step2 {i}")
        result.fetchall()
        print(f"step3 {i}")
        conn.commit()

async def main():
    if os.path.exists("file.db"):
        os.remove("file.db")
    engine = create_engine("sqlite:///file.db")
    await setup(engine)
    await asyncio.gather(
        *[work(engine, i) for i in range(10)]
    )


asyncio.run(main())
```


in the output we see each request is run fully in sequence, one after the other. the two sleep calls do not get awaited until the end, and this is in all cases. The program output is completely deterministic:

```
step1 0
step2 0
step3 0
step1 1
step2 1
step3 1
starting sleep 2
step1 3
step2 3
step3 3
step1 4
step2 4
step3 4
starting sleep 5
step1 6
step2 6
step3 6
step1 7
step2 7
step3 7
step1 8
step2 8
step3 8
step1 9
step2 9
step3 9
ending sleep 2
ending sleep 5
```

switching to use the async API:

```py
import os
import asyncio
from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine

async def setup(engine):
    async with engine.connect() as conn:
        await conn.exec_driver_sql("""
create table big_table (
    some_col int
)
        """)
        await conn.exec_driver_sql(
            "insert into big_table (some_col) values (?)",
            [(i % 100, ) for i in range(1000000)]
        )
        await conn.commit()


async def work(engine, i):
    if i in (2, 5):
        print(f"starting sleep {i}")
        await asyncio.sleep(.1)
        print(f"ending sleep {i}")
        return

    async with engine.connect() as conn:
        print(f"step1 {i}")
        result = await conn.stream(text("select count(*) from big_table group by some_col"))
        print(f"step2 {i}")
        await result.fetchall()
        print(f"step3 {i}")
        await conn.commit()

async def main():
    if os.path.exists("file.db"):
        os.remove("file.db")
    engine = create_async_engine("sqlite+aiosqlite:///file.db")
    await setup(engine)
    await asyncio.gather(
        *[work(engine, i) for i in range(10)]
    )


asyncio.run(main())
```

now the requests are interleaved and take place in a random sequence, because the await keyword is yielding. the two sleep calls finish at a random spot in the output as well:


```py
starting sleep 2
starting sleep 5
step1 8
step1 1
step1 3
step1 7
step1 6
step1 9
step1 0
step1 4
step2 3
ending sleep 2
step2 7
ending sleep 5
step2 1
step2 0
step2 9
step2 8
step2 6
step2 4
step3 7
step3 3
step3 1
step3 0
step3 8
step3 9
step3 6
step3 4

```

we get the identical output with a pure aiosqlite program:

```py
import os
import asyncio
import aiosqlite

async def setup():
    async with aiosqlite.connect("file.db") as conn:
        cursor = await conn.cursor()
        await cursor.execute("""
create table big_table (
    some_col int
)
        """)
        await cursor.executemany(
            "insert into big_table (some_col) values (?)",
            [(i % 100, ) for i in range(1000000)]
        )
        await cursor.close()
        await conn.commit()


async def work(i):
    if i in (2, 5):
        print(f"starting sleep {i}")
        await asyncio.sleep(.1)
        print(f"ending sleep {i}")
        return

    async with aiosqlite.connect("file.db") as conn:
        cursor = await conn.cursor()
        print(f"step1 {i}")
        await cursor.execute("select count(*) from big_table group by some_col")
        print(f"step2 {i}")
        await cursor.fetchall()
        print(f"step3 {i}")
        await cursor.close()
        await conn.commit()

async def main():
    if os.path.exists("file.db"):
        os.remove("file.db")
    await setup()
    await asyncio.gather(
        *[work(i) for i in range(10)]
    )


asyncio.run(main())
```
output:

```
starting sleep 2
starting sleep 5
step1 0
step1 4
step1 6
step1 1
step1 3
step1 9
step1 8
step1 7
step2 0
step2 7
step2 6
step2 4
step2 1
ending sleep 2
step2 8
ending sleep 5
step2 9
step2 3
step3 0
step3 4
step3 6
step3 8
step3 7
step3 1
step3 9
step3 3

```


I can also share with you the [code](https://github.com/zzzeek/async_demo/) used in [this talk I just did for the openstack technical committee](https://youtu.be/KnTDP7Dqit8?si=XEACLA-AwX2i0wYx&t=1900) which illustrates a live walkthrough of SQLAlchemy's greenlet API, where you can watch a `sys.settrace()` example jump between two separate greenlets.

So this is not to say there's not a problem somewhere but you can't say "greenlet is blocking the event loop" without a clear demonstration, this would be an extremely obvious problem with SQLAlchemy's asyncio approach that would have been noticed years ago.

**Reply by msukmanowsky** [reply]
> So this is not to say there's not a problem somewhere but you can't say "greenlet is blocking the event loop" without a clear demonstration, this would be an extremely obvious problem with SQLAlchemy's asyncio approach that would have been noticed years ago.

Fair, but something definitely is going on that seems/feels like FastAPI's loop is being blocked to the point where all requests are experiencing latencies. I'll try also opening up a discussion with the FastAPI team to see if they have any thoughts here and will think on this a bit more to see if I've got any other ideas.

**Reply by zzzeek** [reply]
look at our FastAPI tag added here.    I would **love** for FastAPI people to get involved with these.  We are being crushed with issues that arise from whatever it is they do

**Reply by CaselIT** [reply]
Should have done this sooner. Looked at what you use to test this, and I think the issue is that your test is just wrong. Try running `await run_test(args.num_requests, ["/ping"]*5)` and you will see the same behavior without any query.

**Reply by msukmanowsky** [reply]
> Should have done this sooner. Looked at what you use to test this, and I think the issue is that your test is just wrong. Try running await run_test(args.num_requests, ["/ping"]*5) and you will see the same behavior without any query.

@CaselIT that's taken care of [here](https://github.com/msukmanowsky/fastapi-sqlalchemy-slowness/blob/main/run_test.py#L27-L28). By default, `run_test.py` will issue 100 concurrent requests to each endpoint passed to it. So

```python
await run_test(args.num_requests, ["/ping", "/ping-with-async-sleep"])
```

Will send 100 requests to `/ping` and 100 to `/ping-with-async-sleep` and then afterwards, [gather them](https://github.com/msukmanowsky/fastapi-sqlalchemy-slowness/blob/main/run_test.py#L29).

**Reply by CaselIT** [reply]
sure but that will issue 200 concurrent request. if you just call `['/ping']` it will issue 100 concurrent requests. that's quite a bit of difference

**Reply by msukmanowsky** [reply]
> sure but that will issue 200 concurrent request. if you just call `['/ping']` it will issue 100 concurrent requests. that's quite a bit of difference

I'm not sure I follow your concern, sorry. You're concerned about 200 being too high a volume?

**Reply by CaselIT** [reply]
it's not representative if you test something doing 100 concurrent request or 200 concurrent requests.
to have meaningful number the total number of request should stay the same.

You can simply try doing `await run_test(args.num_requests, ["/ping"]*5)` and you should see times that are significantly slower than running 100 requests only

**Comment by CaselIT** [comment]
> Would appreciate an example showing SQLAlchemy doesn't block the asyncio event loop in anyway as that may spur me to open a separate issue with the FastAPI team.

here you go
```py
import argparse
import asyncio
import time

import sqlalchemy as sa
from sqlalchemy.ext.asyncio import create_async_engine

engine = create_async_engine(
    "postgresql+asyncpg://scott:tiger@127.0.0.1:5432/test", pool_size=100
)


async def count(num):
    value = 0
    start = time.monotonic()
    while True:
        await asyncio.sleep(0.001)
        value += 1
        if value % 200 == 0:
            print("count", num, value, time.monotonic() - start)
            start = time.monotonic()


async def query(num):
    stmt = sa.select(sa.func.pg_sleep(0.3))
    value = 0
    start = time.monotonic()
    while True:
        async with engine.connect() as c:
            await c.execute(stmt)
        value += 1
        if value % 10 == 0:
            print("query", num, value, time.monotonic() - start)
            start = time.monotonic()


async def main(counts, queries):
    await asyncio.gather(
        *(count(i + 1) for i in range(counts)),
        *(query(i + 1) for i in range(queries)),
    )


if __name__ == "__main__":
    a = argparse.ArgumentParser()
    a.add_argument("--counts", type=int, default=10)
    a.add_argument("--query", type=int, default=10)

    args = a.parse_args()
    try:
        asyncio.run(main(args.counts, args.query))
    except BaseException:
        pass
```
